packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -749,34 +749,73 @@ export class Pipeline extends PipelineBase {
+      stageIndex: number;
+      stageName: string;
+      action: FullActionDescriptor;
+    }
+
+    /**
+     * Returns whether a is before or the same order as b
+     */
+    function beforeOrEqual(a: PipelineLocation, b: PipelineLocation) {
+      if (a.stageIndex !== b.stageIndex) { return a.stageIndex < b.stageIndex; }
+      return a.action.runOrder <= b.action.runOrder;
+    }
+
+    function renderLoc(loc: PipelineLocation) {
+      // runOrders are 1-based, so make the stageIndex also 1-based otherwise it's going to be confusing.
+      return `stage ${loc.stageIndex + 1} ('${loc.stageName}') action ${loc.action.runOrder} ('${loc.action.actionName}')`;
+    }
+
+    const producer: Record<string, PipelineLocation> = {};
+    const firstConsumer: Record<string, PipelineLocation> = {};
+
+    for (const [stageIndex, stage] of enumerate(this._stages)) {
+      // For every output artifact, get the producer
+      for (const action of stage.actionDescriptors) {
+        const thisLocation = { stageIndex, action, stageName: stage.stageName };
+
+        for (const outputArtifact of action.outputs) {
+          // output Artifacts always have a name set
+          const name = outputArtifact.artifactName!;
+          if (producer[name]) {
+            ret.push(`Artifact '${name}' is used as an output by both '${producer[name].action.actionName}' and '${action.actionName}'. Every artifact can only be produced once.`);
+            continue;
           }
+
+          producer[name] = thisLocation;
         }
 
-        // then process outputs by adding them to the Set
-        const outputArtifacts = action.outputs;
-        for (const outputArtifact of outputArtifacts) {
-          // output Artifacts always have a name set
-          if (outputArtifactNames.has(outputArtifact.artifactName!)) {
-            ret.push(`Artifact '${outputArtifact.artifactName}' has been used as an output more than once`);
-          } else {
-            outputArtifactNames.add(outputArtifact.artifactName!);
+        // For every input artifact, get the first consumer
+        for (const inputArtifact of action.inputs) {
+          const name = inputArtifact.artifactName;
+          if (!name) {
+            ret.push(`Action '${action.actionName}' has an unnamed input Artifact (probably not used as an output in this pipeline)`);
+            continue;
+          }
+
+          if (!firstConsumer[name] || beforeOrEqual(thisLocation, firstConsumer[name])) {
+            firstConsumer[name] = thisLocation;
           }
         }
       }
     }
 
+    // Now validate that every input artifact is produced before it's
+    // being consumed.
+    for (const [artifactName, consumerLoc] of Object.entries(firstConsumer)) {
+      const producerLoc = producer[artifactName];
+      if (!producerLoc) {
+        ret.push(`Artifact '${artifactName}' is used as an input by '${consumerLoc.action.actionName}', but is not being produced in this pipeline`);
```

Comment:
This phrasing is a little weird... perhaps something like `Artifact '${artifactName}' is used as an input by '${consumerLoc.action.actionName}', but is never used as the output of any action in the pipeline`?

Reply:
So one of these (forgot which) was the one that threw me off.

Both this and the previous error message you commented on actually have the same root cause (input is not being produced), but the error messages are confusingly different. The only difference is one has a name and the other one doesn't, but the user error is exactly the same.

Also "being used as an output" kind implies some optionality to me,  even though there isn't really, and sounds very mechanical. I'd much prefer "is not being produced": it's very clear that you can't consume something that's not being produced.